File: Scripts/Combat/DeckManager.cs
using Roguelike.Data;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Roguelike.Logic
{
    /// <summary>
    /// Manages the state of the hero's deck throughout a run and during combat.
    /// Handles card piles, drawing, shuffling, and modifications to the master deck.
    /// </summary>
    public class DeckManager
    {
        private readonly Random rng;

        /// <summary>
        /// The complete list of cards the hero owns for the entire run
        /// </summary>
        public List<CardData> MasterDeck { get; private set; } = new List<CardData>();

        public List<CardData> DrawPile { get; private set; } = new List<CardData>();
        public List<CardData> Hand { get; private set; } = new List<CardData>();
        public List<CardData> DiscardPile { get; private set; } = new List<CardData>();
        public List<CardData> ExhaustPile { get; private set; } = new List<CardData>();

        public DeckManager(Random randomGenerator)
        {
            rng = randomGenerator;
        }

        /// <summary>
        /// Populates the MasterDeck based on a list of card IDs and the CardPool.
        /// </summary>
        public void InitializeMasterDeck(IEnumerable<string> cardIds, CardPool pool)
        {
            MasterDeck.Clear();
            foreach (var id in cardIds)
            {
                var card = pool.GetCard(id);
                if (card != null)
                {
                    MasterDeck.Add(card);
                }
            }
        }

        /// <summary>
        /// Adds a card to the master deck permanently. Used for rewards.
        /// </summary>
        public void AddCardToMasterDeck(CardData card) => MasterDeck.Add(card);

        /// <summary>
        /// Removes a card from the master deck permanently. Used for shop/event removals.
        /// </summary>
        public void RemoveCardFromMasterDeck(CardData card) => MasterDeck.Remove(card);

        /// <summary>
        /// Prepares the deck for a new combat encounter.
        /// </summary>
        public void StartCombat()
        {
            DrawPile.Clear();
            Hand.Clear();
            DiscardPile.Clear();
            ExhaustPile.Clear();

            DrawPile.AddRange(MasterDeck);
            Shuffle(DrawPile);
        }

        /// <summary>
        /// Draws a specified number of cards into the hand.
        /// </summary>
        public void DrawCards(int amount)
        {
            for (int i = 0; i < amount; i++)
            {
                if (DrawPile.Count == 0)
                {
                    if (DiscardPile.Count == 0)
                    {
                        break;
                    }
                    ReshuffleDiscardIntoDraw();
                }

                var cardToDraw = DrawPile[0];
                DrawPile.RemoveAt(0);
                Hand.Add(cardToDraw);
            }
        }

        /// <summary>
        /// Moves a card from the hand to the discard pile.
        /// </summary>
        public void DiscardCardFromHand(CardData card)
        {
            if (Hand.Remove(card))
            {
                DiscardPile.Add(card);
            }
        }

        /// <summary>
        /// Moves all cards from the hand to the discard pile.
        /// </summary>
        public void DiscardHand()
        {
            DiscardPile.AddRange(Hand);
            Hand.Clear();
        }

        /// <summary>
        /// Moves all cards from the discard pile into the draw pile and shuffles it.
        /// </summary>
        public void ReshuffleDiscardIntoDraw()
        {
            DrawPile.AddRange(DiscardPile);
            DiscardPile.Clear();
            Shuffle(DrawPile);
        }

        /// <summary>
        /// Shuffles a list of cards using the Fisher-Yates algorithm.
        /// </summary>
        private void Shuffle(List<CardData> list)
        {
            int n = list.Count;
            while (n > 1)
            {
                n--;
                int k = rng.Next(n + 1);
                CardData value = list[k];
                list[k] = list[n];
                list[n] = value;
            }
        }
    }
}

File: python_roguelike/core/deck_manager.py
import random
from typing import List, Optional
from .combatant import Combatant
from ..data.card_data import CardData
from ..data.pools.card_pool import CardPool


class DeckManager:
    def __init__(self, rng: random.Random):
        self._rng = rng
        self.master_deck: List[CardData] = []
        self.draw_pile: List[CardData] = []
        self.hand: List[CardData] = []
        self.discard_pile: List[CardData] = []
        self.exhaust_pile: List[CardData] = []

    def initialize_master_deck(self, card_ids: List[str], pool: CardPool):
        self.master_deck.clear()
        for cid in card_ids:
            card = pool.get_card(cid)
            if card is not None:
                self.master_deck.append(card)

    def add_card_to_master_deck(self, card: CardData):
        self.master_deck.append(card)

    def remove_card_from_master_deck(self, card: CardData):
        if card in self.master_deck:
            self.master_deck.remove(card)

    def start_combat(self):
        self.draw_pile.clear()
        self.hand.clear()
        self.discard_pile.clear()
        self.exhaust_pile.clear()
        self.draw_pile.extend(self.master_deck)
        self._shuffle(self.draw_pile)

    def draw_cards(self, amount: int):
        for _ in range(amount):
            if not self.draw_pile:
                if not self.discard_pile:
                    break
                self._reshuffle_discard_into_draw()
            if not self.draw_pile:
                break
            card = self.draw_pile.pop(0)
            self.hand.append(card)

    def discard_card_from_hand(self, card: CardData):
        if card in self.hand:
            self.hand.remove(card)
            self.discard_pile.append(card)

    def discard_hand(self):
        self.discard_pile.extend(self.hand)
        self.hand.clear()

    def _reshuffle_discard_into_draw(self):
        self.draw_pile.extend(self.discard_pile)
        self.discard_pile.clear()
        self._shuffle(self.draw_pile)

    def _shuffle(self, lst: List[CardData]):
        n = len(lst)
        while n > 1:
            n -= 1
            k = self._rng.randint(0, n)
            lst[k], lst[n] = lst[n], lst[k]
